### ISSUE TITLE
Update scripting of Python installer for Python 3.5

### ISSUE THREAD

**Issue by takluyver** [takluyver/pynsist#35]
Python 3.5 will have a rebuilt Python installer, and it seems likely that we will need to pass some slightly different options to it. The details aren't finalised yet, I'll get back to this when they are.

http://stevedower.id.au/blog/the-python-3-5-installer/

**Comment by takluyver** [comment]
The options for the new installer (still in alpha) are here: https://docs.python.org/3.5/using/windows.html#installing-without-ui

Hopefully all we'll need to pass is `/passive`

**Comment by zooba** [comment]
`/passive` should be fine, but I'm experimenting with a self-extracting ZIP of an embeddable Python interpreter. Check out [python-3.5.0a4-embed-win32.exe](https://www.python.org/ftp/python/3.5.0/python-3.5.0a4-embed-win32.exe) and [python-3.5.0a4-embed-amd64.exe](https://www.python.org/ftp/python/3.5.0/python-3.5.0a4-embed-amd64.exe) and let me know (either here, python-dev or directly) whether they meet your needs. pynsist is one of the key use cases for these, so I will happily make whatever changes you need.

**Comment by zooba** [comment]
Also, I've got [issue23955](http://bugs.python.org/issue23955) open to make it easier to isolate an applocal copy of Python 3.5. Currently it looks like we'll recommend adding (rather than providing) a pyvenv.cfg file with a new setting to avoid the registry lookups that occur at startup, so any feedback on how you would make that work for pynsist is also welcome.

**Comment by zooba** [comment]
Oh, you'll also need to pull in the VC14 runtime separately, since that isn't part of the embedded installer. I don't think there's any good place to get that right now apart from installing VS 2015 CTP 6, but eventually it will be installed everywhere through Windows Update and will also have a public download.

**Comment by takluyver** [comment]
Excellent, thanks.

My first priority for Pynsist will be adding an option to use your new network-based installers, which should be pretty simple.

I hope that the current approach of installing Python as a separate framework rather than bundled in the application will work well for most cases - bugfix releases of Python shouldn't usually break things, and the user has a real Python environment that they can easily install extra pieces into, which would be important if we distributed IPython like this, for instance. Shrinking the download size with the web installers will also help with how 'lightweight' users perceive the applications to be, though I'm impressed that you've got the embedded archives down to 6-7MB.

If 'app local' Python builds become a thing, I would definitely want to support them as an option, however, so developers could choose to ship a completely self-contained application.

Fiddling with those self-extracting archives - what format are they? `unzip` doesn't work at all, and `7z` can list the files but not unpack the data... ah, wait, I've got it now, it's RAR. From my point of view, I'd like a regular .zip file that I can easily unpack on any system, even using the Python stdlib. To bundle app-local Python, I'm going to unpack it and repack anyway, so I don't particularly need great compression. And there's no particular need to make it self extracting - I'm going to script the extraction anyway, so the .exe just slightly obscures the format.

Out of interest, what did you cut out to bring the size down? I see that tkinter is missing, which I'd want to make clear in the Pynsist docs. Are they missing anything else that might be relevant?

Aside: is it simple to add an installer option for the start menu shortcuts? See #1. It's not high priority, but it would be nice to be able to install Python like a framework without putting things in the start menu.

**Comment by zooba** [comment]
Yeah, I ended up using WinRAR to create the extractors because they have the most flexible licensing (7-zip's extractor is probably LGPL, WinRAR's is explicitly neutral). I should be able to make it a plain ZIP format, but by making it an extractor it means I can sign the release (which breaks the ability to treat the exe as a zip file anyway).

My expectation was that you could simply run the extractor and provide the output directory (IIRC, "-o <dir>") rather than repackaging the contents. I know some people are keen to install/extract Python for Windows on non-Windows operating systems, but I frankly don't understand or care much about that case. Dealing with one platform is hard enough...

It's smaller because I've omitted the test suite, the documentation, and Tcl/Tk. Possibly the latter would be missed, but I've never seen an app host Python and use it - they always seem to go for PyQt/PySide or wxPython. It wouldn't be too hard to bundle up tkinter separately anyway (I'd personally prefer to see it on PyPI rather than in the stdlib, but that's not really a fight worth fighting right now :) )

Should be easy enough to add a command-line option for the shortcuts, though my concern there is that the "framework" install would conflict with a Python "application" install (consider the cartesian product of each optional feature in two dot versions of Python, the two choices for per-user/all-users and the two choices for installation ordering... gets messy very quickly!). Having the applocal copy of Python is the safest way to make things work, though it does mean that apps are responsible for their own updates.

**Comment by takluyver** [comment]
I'd prefer to unpack at build time rather than install time - as I mentioned on the Python issue, it's easier to write and test build steps than install steps. Since building on non-Windows platforms is a deliberate feature of Pynsist, I will want to unpack it on different platforms. If it's a zip, that's trivial. If it's a self-extracting signed RAR, Pynsist will most likely grow code to unpack it in different ways on different platforms, and require extra non-Python dependencies. I don't know how to check the signature anyway, and the end user won't ever see the Python archive, whatever format it's in.

TLDR: I would still prefer zip files. ;-) Self-extracting RARs offer no advantages for me, only extra complexity.

OK, I think Tcl/Tk is the only bit I need to make clear for application developers. I agree that not many people use it and it's probably worth the space saving, but I'm sure someone uses it. I've used it in toy projects a couple of times.

I'm not quite sure what you mean about the conflicts. If you install 'framework' first, running the installer with the default options ('application' install) should add the shortcuts. If you install 'application' first, a later 'framework' install shouldn't remove anything. And if you install two different versions (e.g. 3.4 and 3.5), they'll be in different locations and not conflicting.

**Comment by zooba** [comment]
Okay, I see the point about the ZIP file, especially since you want cross-platform building. I'll consider the impacts of changing to that, but I think it should be okay.

As for the install conflicts, consider these situations:

1. User has installed Python 3.5.2 for all users on their machine
2. pynsist installs Python 3.5.0 for the current user with no shortcuts
3. py.exe (and hence file associations/shebangs) will prefer per-user 3.5.0 but shortcuts will go to 3.5.2

1. pynsist installs Python 3.5.0 per-user with Tcl/tk
2. User installs Python 3.5.1 per-user without Tcl/tk
3. pynsist app is now missing a dependency

1. User installs Python 3.5.0 with documentation
2. pynsist installs Python 3.5.1 without documentation
3. User loses documentation

And there are plenty more. The logic to handle all of them is basically impossible to sort out completely and there are always going to be edge cases. It's my current opinion that users should be completely aware of the EXE installer and completely unaware of the ZIP extractor, as this is the only way to maintain sanity.

**Comment by takluyver** [comment]
Great, thanks. :-)

OK, that makes sense. Once I've got app-local Python working, I'll consider switching it to be the default.

**Comment by takluyver** [comment]
@zooba , I'm working on supporting applocal Python for Pynsist installers. I've got a simple example installer building, but when I try to use the embedded Python install (this is on Win 7), I get an error saying:

> The program can't start because VCRUNTIME140.dll is missing from your computer. Try reinstalling the program to fix this problem.

I guess that means I need the MSVC runtime, but I thought the point of the embeddable builds was that they were entirely self contained, and didn't rely on anything being installed externally. Shouldn't they include the runtime? Or do I need to work out a way to install that separately?

**Comment by zooba** [comment]
I think it's best for you to work out a way to install it separately. The hope is that the bundle will be predominantly used for embedding in other apps (as opposed to bundling), and so the CRT will be a requirement anyway. I can't redistribute the CRT files directly as they are an operating system component, and because they're not officially released they haven't been pushed out through Windows Update yet either.

If you grab the current Python 3.5 installer and run it with `/layout [directory]` it will download all the dependencies, which includes a set of `.msu` files. These have the RC version of the runtime (currently the latest publicly available version) for each target version of Windows. Running these with `wusa.exe` on the target system is the "correct" way to install just the C runtime, and is the direction I've chosen to go with CPython. However, this doesn't include `vcruntime140.dll`.

`vcruntime140.dll` *can* be redistributed, but I hesitate to include it in the zipped bundle because it's more likely to conflict than help in the embedding case. Same goes for statically linking the entire interpreter (not to mention the duplicated code would make it huge...). Easiest to extract that from a Python 3.5 install, or if you've got VS 2015 then it's under `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\x86\Microsoft.VC140.CRT` (or `x64\Microsoft.VC140.CRT`).

Finally, the easiest and most official way to install the CRT is to use the full installer. These are in `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\1033`, but aren't available for separate download **yet** because VS 2015 hasn't released. The downside of these is that they are about 13MB each (though you only need the 32-bit or 64-bit to match your interpreter, not to match the target system). Eventually they'll be available online so you can detect and lazily download them, but right now there's no better option.

**Comment by zooba** [comment]
> I thought the point of the embeddable builds was that they were entirely self contained, and didn't rely on anything being installed externally

That's only part of it. The real benefit is that they aren't affected by anything being installed externally (e.g. PYTHONPATH, registry keys, user site-packages, etc.).

**Comment by zooba** [comment]
Just heard that I can statically link `vcruntime140.dll` without the rest of the CRT, so I'm going to investigate doing that with the next release. That way you'll only need to redistribute the `.msu` files (and I would hope there's a way to force Windows Update to download and install them itself, but I don't know what that is right now).

**Comment by takluyver** [comment]
OK, I'll wait and see what transpires with that option, then.

You mentioned that the runtime will be pushed out through Windows update. Which versions of Windows will that reach, and will it be pushed out in a way that will automatically get installed with the default update settings?

**Comment by zooba** [comment]
Should go all the way back to Vista, but it may not be required that far back. Probably depends on what other updates require it.

**Comment by takluyver** [comment]
The error message is a bit different with beta 3:

> The program can't start because api-ms-win-crt-runtime-l1-1-0.dll is missing from your computer. Try reinstalling the program to fix this problem. 

Does that mean you did the static linking? Can I include that DLL with the application?

**Comment by zooba** [comment]
Yeah, I added the static linking of `vcruntime140.dll`, so now all of the CRT should be installed via Windows Update. I expect the CRT redistributables will be available for separate download next week, so the cheapest way for you to handle this will be to detect `ucrtbase.dll` on the machine and then automatically download/install the CRT. It's an operating system component now, and not redistributable apart from the EXE or MSU packages.

**Comment by takluyver** [comment]
Thanks!

What's the best way to check for `ucrtbase.dll` at install time? Sorry if this should be obvious; I know little about Windows.

Did you have any luck with a way to force Windows Update to download and install a specific update on demand? Or will I need to package the MSU in the installer and use wusa to install it?

Also, looking at the docs for wusa, it can trigger a system restart. Does installing the CRT require a restart? Does it depend on other conditions? If it needs a restart, I could suppress that with the `/norestart` flag and trigger a restart when the installer is finished.

**Comment by zooba** [comment]
`LoadLibrary` is probably sufficient for checking for `ucrtbase.dll`, or check out [`IsCrtInstalled()`](https://hg.python.org/cpython/file/361d7af9396e/Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp#l2545) in the CPython installer (though I also needed a version check there to ensure prerelease versions are upgraded).

I still don't know of any way other than packaging the MSU, but if you can do a download on demand I'd strongly recommend that, or else you'll have to include 8 different MSUs and choose at install time.

Typically installing the CRT won't require a restart, but if there is an older version installed and it's in use while the update is applied then it will. Suppressing it initially and prompting at the end of your installer is the best way to deal with that, though it's possible that new processes will get the new CRT without a restart - I'm not entirely sure how the update works, but possibly it's just to make sure the other process gets restarted with the new CRT. Though I could be wrong.

**Comment by takluyver** [comment]
I looked into download on demand for Python, and steered clear of it because it wasn't clear that any NSIS tools could use HTTPS and do certificate validation. I also try to keep as much logic as possible in the build step, because it's easier to write and test Python code than NSI scripts.

Why are there 8 different MSUs? How big will each be?

Is there a way to check whether a restart is required? I don't want to make most users restart when there's really no need, but I also don't want to leave the users who do need a restart with nothing but confusing error messages.

**Comment by zooba** [comment]
Windows Vista, 7, 8 and 8.1, 32-bit and 64-bit. Because it's an OS component, there's a different version for each OS (for Windows XP you can redist the file directly, since it won't get security updates anyway, but Python 3.5 doesn't work on XP anyway so you probably don't care).

The 32-bit files are about 500kb each and the 64-bit files are about 1MB each (since 64-bit OSs need both 32-bit and 64-bit files), so you're looking at an extra 6MB.

As for the restart, if the exit code from `wusa` is `0x00240005` (found in `C:\Program Files (x86)\Windows Kits\8.1\Include\um\wuerror.h`) then you need to reboot (this is an HRESULT, so it's only an error if 0x80000000 is set). There may be some other relevant codes in that file too, but in general if the exit code is positive then it installed okay.

**Comment by takluyver** [comment]
Is that 32 and 64 bit for the things that run on them, or for the OSs themselves? I thought from Vista onwards it was all 64 bit platforms?

That presumably means I need some logic to detect which platform it is at install time to select the correct update. Myeh, that won't be fun, but it looks like there are ways to do it in NSIS.

An idea occurred to me with downloading on demand: are MSU files signed, and does wusa.exe verify a signature? If I can rely on that, then it's OK to download them over insecure HTTP, which was what stopped me from downloading Python on demand.

Thanks for the exit code info, I was hoping that there would be something like that.

**Comment by zooba** [comment]
It's for the OS itself. Being an OS update, it's expected that the OS is the one to do the download, so it'll be able to get the right version. For the CPython web installer, that's exactly what I do, but the offline installer includes all of them.

MSU files are signed, and I assume that wusa will validate the signature, possibly even against a specific certificate, since nobody other than Microsoft should be releasing these updates.

**Comment by zooba** [comment]
Just changed a random byte in one of the files and it refused to install (error 0x8007000d Invalid Data) and the certificate now shows as invalid. Could be some other non-certificate check going on there, but there's some sort of verification.

**Comment by takluyver** [comment]
OK, I think I can work with that once the files are released.

Is there some way you can double check that wusa.exe does check signatures? I agree with you that it seems most likely that it does, but for security related stuff I'd like to have something a bit more definite. The 'invalid data' error could easily be a straightforward checksum for data integrity.

**Comment by takluyver** [comment]
Are the MSUs available now?

Also, I can't see any NSIS commands that will search for a DLL in the same order as LoadLibrary. Can I check for ucrtbase.dll in one of the known directories that NSIS provides as a constant, like $SYSDIR or $WINDIR?

**Comment by zooba** [comment]
Checking for ucrtbase.dll in $SYSDIR (presumably that's going to be `C:\Windows\System32` in most cases) is probably the best you can do. There may be some issues with people who had prerelease versions installed and haven't run updates since it was officially released, but most people who have it should be up to date. You can also skip installing it for Windows 10 as it is part of the default install.

The easiest way to get the MSUs is to download the [redistributables](http://www.microsoft.com/en-us/download/details.aspx?id=48145), [WiX](http://wixtoolset.org) (just the binaries will be sufficient) and run `dark.exe vcredist_x64.exe -x <output directory>`. Then you'll find them deep under the `AttachedContainer` directory.

**Comment by takluyver** [comment]
Thanks, I'll see what I can do with that.

I'm going to work for now assuming that wusa checks tħe signatures on the MSUs, so it's safe to download them over an insecure connection. Can you find someone who can confirm that that's the case? If it's not, I'll have to rethink how it works. Thanks.

**Comment by zooba** [comment]
Had a look and didn't find anyone obvious to ask or an easy way to invalidate the certificate without also invalidating the entire file. You could try running `wusa /log:log.evtx <path to MSU>` and seeing whether there are any messages about signing, or maybe extract the certificate and add it to the revocation list before installing it (on a new machine, presumably). Otherwise, download over SSL or bundling with the app may be options.

**Comment by takluyver** [comment]
OK, I'll try with the `/log` option when I'm at my Windows VM.

**Comment by takluyver** [comment]
Just to check I'm on the right page: the x86 installer and the x64 installer from the MS download page you linked to *both* contain a full copy of all the x86 and x64 MSUs, and those msu files from each are completely identical. Right?

**Comment by takluyver** [comment]
OK, after quite a bit of fiddling, I have something that works on my Windows 7 VM. Can you give [this sample installer](https://www.dropbox.com/s/yzepu548xjt6ldf/Guess_the_Number_1.0.exe?dl=1) a try on any handy systems with different versions of Windows?

Using Python 3.5b4, the installer for a plain console program weighs in at 7.5MB. If it doesn't find `$SYSDIR\ucrtbase.dll` present, it downloads one of eight MSUs I extracted from the VC++ redistributable, and invokes wusa to install it.

**Comment by zooba** [comment]
Worked nicely on Windows 10 (as I'd expect, since the CRT is already installed). Haven't had a chance to try it with other OSs, but I've pinged some people who will probably use 8.1 to try it out.

**Comment by takluyver** [comment]
Thanks. If it works for them, I'll release v 1.6. I'm documenting the support for bundled Python as experimental for now.

**Comment by zooba** [comment]
Looks good. You may want to hold the release until next week though, as we're releasing the first release candidate this weekend.

Also FYI, in rc1 the embedded stdlib will be precompiled .pyc files rather than original sources. In theory it should just be a performance benefit, but there may be obscure errors come out of it. You're likely to be the first to notice :)

**Comment by takluyver** [comment]
OK, great. I'll build that installer again with RC1 and get you to check it
still works before I release, then. Thanks!

**Comment by zooba** [comment]
Just discovered http://www.microsoft.com/en-us/download/details.aspx?id=48234, which provides the UCRT update packages in a ZIP file. They should be identical to what we extracted earlier.

**Comment by takluyver** [comment]
Thanks. For now, I've put the MSU files in a branch of this repo: https://github.com/takluyver/pynsist/tree/msvcrt-2015-msus . It doesn't seem like an ideal place for them, but I couldn't think of a better one.

Here's the same example installer as before built with 3.5rc1:
https://www.dropbox.com/s/da44syi1ywke1fd/Guess_the_Number_1.0_py35rc1.exe?dl=1

**Comment by zooba** [comment]
Works great (though I only tried on Win10, so no need for the updates).

Would be nice for the Uninstall link to include the program name, as that helps people find it when searching.

**Comment by takluyver** [comment]
What do you mean about the uninstall link? Does Windows 10 change the interface of the add/remove programs thing?

**Comment by zooba** [comment]
The link in the start menu is just called "Uninstall", and when you search it shows up without the context of the group it's in.

This is why we changed Python recently to install as "Python 3.5 (32-bit)" rather than "Python (command line)" - the latter does not tell you which version you are looking at when you have just searched for "python".

**Comment by takluyver** [comment]
The installers pynsist creates shouldn't be adding an uninstall item in the start menu at all - it should add the registry entries so the application appears in 'add/remove programs'.

**Comment by zooba** [comment]
It did both for me.

**Comment by takluyver** [comment]
OK, I'm baffled by that, but I've made a new issue (#54) to investigate it. If you have some idea for how to debug it, let's go for it, otherwise I'll release 1.6.

**Comment by zooba** [comment]
Strange, I didn't get it just then (on my work machine - last time was one of my laptops).

Maybe it's not a real issue after all... sorry for the noise.

**Comment by takluyver** [comment]
I just remembered that the scripting of the Python installer itself has also changed. If you have a moment, can you test this installer:
https://www.dropbox.com/s/gginq0g0eqnr4bt/My_App_1.0.exe?dl=1

It should run the Python 3.5.0rc1 installer as part of its own install procedure. I know this is problematic and things should move towards the embeddable builds, but this is what Pynsist has done up to now, and I don't want to drop it quite yet.

**Comment by takluyver** [comment]
@zooba I just got to do some testing on a Windows 7 machine, and I had some trouble with the options for the Python installer - `/p` for passive still shows a confirmation screen, whereas I'd like it to just start installing without prompting the user at all. `/q` for quiet didn't seem to work - I could hear the hard drive churning away for a bit, but there was no evidence that Python got installed (and of course, no visible error). I also tried SImpleInstall=1, but there's still a confirmation page. Any suggestions?

**Comment by zooba** [comment]
They may need to be spelled out fully (`/passive` and `/quiet`). You should also have log files in your Temp directory that may have more information.

SimpleInstall is just for the UI. It's not really relevant with any of these options.

**Comment by takluyver** [comment]
OK, `/passive` works where `/p` doesn't - you might want to update [the docs](https://docs.python.org/3.5/using/windows.html#installing-without-ui), which currently say that those are equivalent. `/q` for quiet does appear to be working, though.

The docs also say that the default for InstallAllUsers is 1, whereas in fact it appears to be 0.

However, the actual installation still fails. It creates two log files - here's the relevant bit from the first:

```
[13D4:0E00][2015-09-11T14:18:06]i301: Applying execute package: core_AllUsers, action: Install, path: C:\ProgramData\Package Cache\{F486180D-53F6-4AC5-9B15-2F9D810BC1C0}v3.5.124.0\core.msi, arguments: ' ALLUSERS="1" ARPSYSTEMCOMPONENT="1" MSIFASTINSTALL="7" TARGETDIR="" OPTIONALFEATURESREGISTRYKEY="Software\Python\PythonCore\3.5\InstalledFeatures"'
[13D4:1300][2015-09-11T14:18:06]i304: Verified existing payload: dev_AllUsers at path: C:\ProgramData\Package Cache\{B46934A6-7862-474C-A07F-AEEFD37BCEBA}v3.5.124.0\dev.msi.
[13D4:1300][2015-09-11T14:18:06]i304: Verified existing payload: exe_AllUsers at path: C:\ProgramData\Package Cache\{EC503F3D-4479-4E8A-96D4-D322139F5CFC}v3.5.124.0\exe.msi.
[13D4:0E00][2015-09-11T14:18:06]e000: Error 0x80070643: Failed to install MSI package.
[13D4:0E00][2015-09-11T14:18:06]e000: Error 0x80070643: Failed to execute MSI package.
[1374:04F4][2015-09-11T14:18:06]e000: Error 0x80070643: Failed to configure per-machine MSI package.
[1374:04F4][2015-09-11T14:18:06]i319: Applied execute package: core_AllUsers, result: 0x80070643, restart: None
```

And from the second (`...core_AllUsers.log`):

```
Action start 14:18:06: LaunchConditions.
MSI (s) (A4:44) [14:18:06:646]: Note: 1: 2205 2:  3: Error 
MSI (s) (A4:44) [14:18:06:646]: Note: 1: 2228 2:  3: Error 4: SELECT `Message` FROM `Error` WHERE `Error` = 1709 
MSI (s) (A4:44) [14:18:06:646]: Product: Python 3.5.0rc4 Core Interpreter (64-bit) -- The TARGETDIR variable must be provided when invoking this installer.

The TARGETDIR variable must be provided when invoking this installer.
Action ended 14:18:06: LaunchConditions. Return value 3.
Action ended 14:18:06: INSTALL. Return value 3.
```

It looks like the installer isn't setting a default target directory.

**Comment by takluyver** [comment]
I can send the full log files if you need more information, but I didn't want to post them publicly in case they capture some sensitive information from the system.

**Comment by zooba** [comment]
I think I'll need the full logs. You have/can find my email address, yes?

**Comment by takluyver** [comment]
[Yes](https://github.com/zooba) ;-)

**Comment by zooba** [comment]
I can never tell what's shown publicly and what isn't :)

Looks like an installer bug! If you specify `TargetDir="Path"` it'll be fine (all my tests and automation specify the path), but we're not picking up the default without the UI interaction. I'll file an issue - may be fixed for 3.5.0 but it may slip.

**Comment by takluyver** [comment]
Great, thanks. Calculating the default install path shouldn't be an option. Please add me (takluyver) to the nosy list on the Python issue you file.

**Comment by takluyver** [comment]
(shouldn't be a *problem*, not 'option' - don't know where my mind went then)

**Comment by zooba** [comment]
Issue is at http://bugs.python.org/issue25071

**Comment by takluyver** [comment]
@zooba thanks for all your help with this! I've just released version 1.6 with these changes.

I've also made it set the user-agent header when downloading Python builds, so if you have access to the server logs, you should be able to see how many downloads come from Pynsist. I'm sure it's a small number, but it might be interesting one day.
